File: src/test/java/edu/kit/iti/formal/stvs/model/expressions/SimpleExpressions.java
package edu.kit.iti.formal.stvs.model.expressions;

/**
 * Created by philipp on 17.01.17.
 */
public final class SimpleExpressions {

  private SimpleExpressions() {
  }

  public static Expression negate(Expression e) {
    return new UnaryFunctionExpr(UnaryFunctionExpr.Op.UNARY_MINUS, e);
  }

  public static Expression not(Expression e) {
    return new UnaryFunctionExpr(UnaryFunctionExpr.Op.NOT, e);
  }

  public static Expression and(Expression e1, Expression e2) {
    return new BinaryFunctionExpr(BinaryFunctionExpr.Op.AND, e1, e2);
  }

  public static Expression plus(Expression e1, Expression e2) {
    return new BinaryFunctionExpr(BinaryFunctionExpr.Op.PLUS, e1, e2);
  }

  public static Expression minus(Expression e1, Expression e2) {
    return new BinaryFunctionExpr(BinaryFunctionExpr.Op.PLUS, e1, e2);
  }

  public static Expression equal(Expression e1, Expression e2) {
    return new BinaryFunctionExpr(BinaryFunctionExpr.Op.EQUALS, e1, e2);
  }

  public static Expression lessThan(Expression e1, Expression e2) {
    return new BinaryFunctionExpr(BinaryFunctionExpr.Op.LESS_THAN, e1, e2);
  }

  public static Expression lessEqual(Expression e1, Expression e2) {
    return new BinaryFunctionExpr(BinaryFunctionExpr.Op.LESS_EQUALS, e1, e2);
  }

  public static Expression greaterEqual(Expression e1, Expression e2) {
    return new BinaryFunctionExpr(BinaryFunctionExpr.Op.GREATER_EQUALS, e1, e2);
  }


  public static Expression var(String name) {
    return new VariableExpr(name);
  }

  public static Expression var(String name, int index) {
    return new VariableExpr(name, index);
  }

  public static Expression literal(int integer) {
    return new LiteralExpr(new ValueInt(integer));
  }

  public static Expression literal(boolean bool) {
    return new LiteralExpr(ValueBool.of(bool));
  }

  public static Expression literal(ValueEnum e) {
    return new LiteralExpr(e);
  }

  public static Expression literalEnum(String value, TypeEnum sourceType) {
    return literal(new ValueEnum(value, sourceType));
  }

}

File: stvs/src/test/kotlin/edu/kit/iti/formal/stvs/model/expressions/SimpleExpressions.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.model.expressions

import edu.kit.iti.formal.stvs.model.expressions.ValueBool.Companion.of

/**
 * Created by philipp on 17.01.17.
 */
object SimpleExpressions {
    fun negate(e: Expression): Expression = UnaryFunctionExpr(UnaryFunctionExpr.Op.UNARY_MINUS, e)

    fun not(e: Expression): Expression = UnaryFunctionExpr(UnaryFunctionExpr.Op.NOT, e)

    fun and(e1: Expression, e2: Expression): Expression = BinaryFunctionExpr(BinaryFunctionExpr.Op.AND, e1, e2)

    fun plus(e1: Expression, e2: Expression): Expression = BinaryFunctionExpr(BinaryFunctionExpr.Op.PLUS, e1, e2)

    fun minus(e1: Expression, e2: Expression): Expression = BinaryFunctionExpr(BinaryFunctionExpr.Op.PLUS, e1, e2)

    fun equal(e1: Expression, e2: Expression): Expression = BinaryFunctionExpr(BinaryFunctionExpr.Op.EQUALS, e1, e2)

    fun lessThan(e1: Expression, e2: Expression): Expression =
        BinaryFunctionExpr(BinaryFunctionExpr.Op.LESS_THAN, e1, e2)

    fun lessEqual(e1: Expression, e2: Expression): Expression =
        BinaryFunctionExpr(BinaryFunctionExpr.Op.LESS_EQUALS, e1, e2)

    fun greaterEqual(e1: Expression, e2: Expression): Expression =
        BinaryFunctionExpr(BinaryFunctionExpr.Op.GREATER_EQUALS, e1, e2)

    fun variable(name: String): Expression = VariableExpr(name)

    fun variable(name: String, index: Int): Expression = VariableExpr(name, index)

    fun literal(integer: Int): Expression = LiteralExpr(ValueInt(integer))

    fun literal(bool: Boolean): Expression = LiteralExpr(of(bool))

    fun literal(e: ValueEnum): Expression = LiteralExpr(e)

    fun literalEnum(value: String, sourceType: TypeEnum): Expression = literal(ValueEnum(value, sourceType))
}